### PR TITLE
Added Lua package

### PR DESCRIPTION
**Pull request by Dinno**

### CODE REVIEW

File: cmake/projects/Lua/hunter.cmake [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+include(hunter_download)
+include(hunter_pick_scheme)
+
+hunter_add_version(
+    PACKAGE_NAME
+    Lua
+    VERSION
+    "5.3.2"
+    URL
+    "https://github.com/hunter-packages/lua/archive/5.3.2.tar.gz"
```
Comment (ruslo) [inline cmake/projects/Lua/hunter.cmake]: I've applied some fixes for Linux:
* https://github.com/hunter-packages/lua/commit/54d84610f3a2fda8b0afdfc45d2977c27d788ea9
* https://github.com/hunter-packages/lua/commit/1054f9fa7a86dd15ecdc085430c038883bbdf0ae

Please update archive: https://github.com/hunter-packages/lua/releases/tag/v5.3.2-p0

File: cmake/projects/Lua/hunter.cmake [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    SHA1
+    5395f8637c14b41f950b28f2c1dba65f30fa7030
+)
+
+hunter_pick_scheme(DEFAULT url_sha1_cmake)
+hunter_download(PACKAGE_NAME Lua)
```
Comment (ruslo) [inline cmake/projects/Lua/hunter.cmake]: "No newline at end of file" warning

File: cmake/projects/Lua/hunter.cmake [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    Lua
+    VERSION
+    "5.3.2"
+    URL
+    "https://github.com/hunter-packages/lua/archive/5.3.2.tar.gz"
+    SHA1
+    5395f8637c14b41f950b28f2c1dba65f30fa7030
+)
+
+hunter_pick_scheme(DEFAULT url_sha1_cmake)
```
Comment (ruslo) [inline cmake/projects/Lua/hunter.cmake]: Please add `hunter_cacheable` here

File: cmake/projects/Lua/hunter.cmake [hunk 4]
```diff
@@ -0,0 +1,23 @@
+if(DEFINED HUNTER_CMAKE_PROJECTS_LUA_HUNTER_CMAKE_)
```
Comment (ruslo) [inline cmake/projects/Lua/hunter.cmake]: Feel free to leave a copyrights note

File: cmake/projects/Lua/hunter.cmake [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    PACKAGE_NAME
+    Lua
+    VERSION
+    "5.3.2"
+    URL
+    "https://github.com/hunter-packages/lua/archive/5.3.2.tar.gz"
+    SHA1
+    5395f8637c14b41f950b28f2c1dba65f30fa7030
+)
+
```
Comment (ruslo) [inline cmake/projects/Lua/hunter.cmake]: Probably `hunter_configuration_types(Lua CONFIGURATION_TYPES Release)` need to be added here. By default debug + release variant is built but executable has same name for both (i.e. will be overwritten).

See wiki: https://github.com/ruslo/hunter/wiki/usr.adding.new.package#build-types

File: cmake/projects/Lua/hunter.cmake (hunk 1 above)
Comment (Dinno) [inline cmake/projects/Lua/hunter.cmake]: That's not so easy as it seems :) Many subtle details are revealed when you have to deal with multi-platform development. You make very good work. Thank you!

Comment (ruslo) [inline cmake/projects/Lua/hunter.cmake]: > You make very good work. Thank you!

Enjoy! :)

Comment (ruslo) [inline cmake/projects/Lua/hunter.cmake]: Please apply my Linux fixes by using this code:

```cmake
URL
"https://github.com/hunter-packages/lua/archive/v5.3.2-p0.tar.gz"
SHA1
39fc2bb5e7de0c9b4dd3940ccf99cef3b7cb4b04
```

You can leave version `5.3.2` or update it to `5.3.2-p0` to show that package is patched and not the original.

**Comment by stohrendorf** [comment]
On my system, I need to additionally link to `dl` to make use of this package, because the Lua package makes use of `dlclose`.

**Comment by ruslo** [comment]
> On my system, I need to additionally link to dl to make use of this package, because the Lua package makes use of dlclose

@stohrendorf How did you build it, on what system? Can you show minimal example? I've just found out that some definitions are missing in CMake build which makes Lua not use `dlclose`.

**Comment by stohrendorf** [comment]
Sorry. After updating the Lua package reference to a more recent commit in by Hunter fork, the problem vanished.

**Comment by ruslo** [comment]
@stohrendorf Good, thanks for the information.

**Comment by Dinno** [comment]
I think, that we can merge this pull request when tests will go through.

**Comment by ruslo** [comment]
> I think, that we can merge this pull request when tests will go through

Well, actually there is nothing to test with Lua so test result will not show much info :)

I've added example:
* https://github.com/ruslo/hunter/commit/c6f0556e8de8182bf3a3ef0e8ce10ffd2fa1d502

**Comment by ruslo** [comment]
New wiki page: https://github.com/ruslo/hunter/wiki/pkg.lua

**Comment by ruslo** [comment]
> I've added example

@Dinno Analyze toolchain failed:
```
/.../Lua/Source/src/loslib.c:274:24: warning: Access to field 'tm_sec' results in a dereference of a null pointer (loaded from variable 'stm')
    setfield(L, "sec", stm->tm_sec);
                       ^~~~~~~~~~~
```
Build log: https://travis-ci.org/ruslo/hunter/jobs/102250101
Will remove it for now: https://github.com/ruslo/hunter/commit/da1df8f9812b836281466166f02fcc3fad9a4d48

**Comment by Dinno** [comment]
If stm is NULL it calls luaL_error, which in turn calls lua_error. As we can see from comment in it's code, lua_error never returns, but static analyzer does not know about that, perhaps because it is side effect of calling some kernel function behavior of which is not known to it.
```
LUA_API int lua_error (lua_State *L) {
  lua_lock(L);
  api_checknelems(L, 1);
  luaG_errormsg(L);
  /* code unreachable; will unlock when control actually leaves the kernel */
  return 0;  /* to avoid warnings */
}
```

**Comment by ruslo** [comment]
> I think, that we can solve that with additional return before luaL_error

There is some macro to detect clang analyzer. You can protect additional code with it so it will be obvious why code added.

**Pull request merged**